aten/src/ATen/native/cpu/zmath.h

```diff
@@ -199,32 +200,56 @@ inline std::complex<double> trunc_impl (std::complex<double> z) {
 
 template <typename TYPE>
 inline TYPE max_impl (TYPE a, TYPE b) {
-  return std::max(a, b);
+  if (_isnan<TYPE>(a) || _isnan<TYPE>(b)) {
+    return std::numeric_limits<TYPE>::quiet_NaN();
+  } else {
+    return std::max(a, b);
+  }
 }
 
 template <>
 inline std::complex<float> max_impl (std::complex<float> a, std::complex<float> b) {
-  return std::complex<float>(std::abs(a) > std::abs(b) ? a : b);
+  if (_isnan<std::complex<float>>(a) || _isnan<std::complex<float>>(b)) {
+    return std::complex<float>(std::numeric_limits<float>::quiet_NaN());
+  } else {
+    return std::complex<float>(std::abs(a) > std::abs(b) ? a : b);
+  }
 }
 
 template <>
 inline std::complex<double> max_impl (std::complex<double> a, std::complex<double> b) {
-  return std::complex<double>(std::abs(a) > std::abs(b) ? a : b);
+  if (_isnan<std::complex<double>>(a) || _isnan<std::complex<double>>(b)) {
+    return std::complex<double>(std::numeric_limits<double>::quiet_NaN());
+  } else {
+    return std::complex<double>(std::abs(a) > std::abs(b) ? a : b);
+  }
 }
 
 template <typename TYPE>
 inline TYPE min_impl (TYPE a, TYPE b) {
-  return std::min(a, b);
+  if (_isnan<TYPE>(a) || _isnan<TYPE>(b)) {
```

Comment:
It's better to have min_impl and max_impl as follows (this way you don't have to have separate instantiations for complex float and double):
```
template <typename TYPE, std::enable_if_t<!c10::is_complex_t<TYPE>::value, int> =0>
inline TYPE min_impl(TYPE a, TYPE b) {
    return std::min(a, b);
}

template <typename TYPE, std::enable_if_t<c10::is_complex_t<TYPE>::value, int> =0>
inline TYPE min_impl (TYPE a, TYPE b) {
    return (std::abs(a) < std::abs(b) ? a : b);
}
```
I'm omitting _isnan branch here for clarity, of course you should have it. Similar approach is used for `_isnan` definition, you can take a look here https://github.com/pytorch/pytorch/blob/9f7708eecb25268702ac107471fd13972fb4f5d2/aten/src/ATen/NumericUtils.h#L16

Reply:
@ngimel, it is a good point, but there has problem, for complex number , if one of them is Nan, how to return a Nan complex number? **std::numeric_limits** should given a data type, float of double.

Reply:
This by itself is not a problem, `return TYPE(std::numeric_limits<double>::quiet_NaN())` would do the right thing. However, now I realize that it does not correctly propagate complex nans - if imaginary part of a or b is nan, a number different from a or b will be returned, with real part equal to nan, and imaginary set to 0. So the complex template should be
```
   if (_isnan(a)) 
      return a;
   else if (_isnan(b))
      return b;
   else ...
```

Reply:
@ngimel , thanks!